File: String/008. String to Integer (atoi).cpp
/*
8. String to Integer (atoi)

Implement atoi which converts a string to an integer.

The function first discards as many whitespace characters as necessary until the first non-whitespace character is found. 
Then, starting from this character, takes an optional initial plus or minus sign followed by as many numerical digits as possible, 
and interprets them as a numerical value.

The string can contain additional characters after those that form the integral number, 
which are ignored and have no effect on the behavior of this function.

If the first sequence of non-whitespace characters in str is not a valid integral number, 
or if no such sequence exists because either str is empty or it contains only whitespace characters, no conversion is performed.

If no valid conversion could be performed, a zero value is returned.

Note:

Only the space character ' ' is considered as whitespace character.
Assume we are dealing with an environment which could only store integers within the 32-bit signed integer range: 
[−231,  231 − 1]. If the numerical value is out of the range of representable values, INT_MAX (231 − 1) or INT_MIN (−231) is returned.
Example 1:

Input: "42"
Output: 42
Example 2:

Input: "   -42"
Output: -42
Explanation: The first non-whitespace character is '-', which is the minus sign.
             Then take as many numerical digits as possible, which gets 42.
Example 3:

Input: "4193 with words"
Output: 4193
Explanation: Conversion stops at digit '3' as the next character is not a numerical digit.
Example 4:

Input: "words and 987"
Output: 0
Explanation: The first non-whitespace character is 'w', which is not a numerical 
             digit or a +/- sign. Therefore no valid conversion could be performed.
Example 5:

Input: "-91283472332"
Output: -2147483648
Explanation: The number "-91283472332" is out of the range of a 32-bit signed integer.
             Thefore INT_MIN (−231) is returned.

*/

class Solution {
public:
    int myAtoi(string str) {
        int i = 0;
        while(i<str.size() && str[i]==' ') i++;
        if(i>=str.size() || str[i]!='-' && str[i]!='+' && !isdigit(str[i])) return 0;
        int sign = 1;
        if(!isdigit(str[i]))
            sign = str[i++] == '-' ? -1 : 1;
        long cur = 0;
        while(i<str.size() && cur<numeric_limits<int>::max() && isdigit(str[i]))
            cur = cur*10 + str[i++]-'0' ;
        if (cur > numeric_limits<int>::max())
            return sign == 1 ? numeric_limits<int>::max() : numeric_limits<int>::min();
        else 
            return cur*sign;
    }
};


static int hello=[](){
    std::ios::sync_with_stdio(false);
    cin.tie(NULL);
    return 0;
}();
class Solution {
public:
    int myAtoi(string str) {
        int temp = 0;
        stringstream stream(str);
        stream>>temp;
        return temp;
    }
};



class Solution {
public:
	inline bool isdigit(char c)
	{
		return c >= '0'&c <= '9';
	}

	int myAtoi(string str) {
		if (str.size() == 0)return 0;
		int i = 0, res = 0;
		bool minus = false;
		while (str[i] == ' ')i++;
		if (str[i] == '-'){
			minus = true;
			++i;
		}
		else if (str[i] == '+')++i;
		else if (!isdigit(str[i]))return 0;

		while (isdigit(str[i])){
            int tmp=res*10+str[i] - '0';
			if (tmp/10!=res)return minus ? INT_MIN : INT_MAX;
            res=tmp;
			++i;
		}
		return minus ? -res : res;
	}
};

class Solution {
public:
    int myAtoi(string str) {
        long result = 0;
        int indicator = 1;
        for(int i = 0; i<str.size();)
        {
            i = str.find_first_not_of(' ');
            if(str[i] == '-' || str[i] == '+')
                indicator = (str[i++] == '-')? -1 : 1;
            while('0'<= str[i] && str[i] <= '9') 
            {
                result = result*10 + (str[i++]-'0');
                if(result*indicator >= INT_MAX) return INT_MAX;
                if(result*indicator <= INT_MIN) return INT_MIN;                
            }
            return result*indicator;
        }
    }
};

File: String/008. String to Integer (atoi).py
"""
8. String to Integer (atoi)

Example 1:

Input: "42"
Output: 42
Example 2:

Input: "   -42"
Output: -42
Explanation: The first non-whitespace character is '-', which is the minus sign.
             Then take as many numerical digits as possible, which gets 42.
Example 3:

Input: "4193 with words"
Output: 4193
Explanation: Conversion stops at digit '3' as the next character is not a numerical digit.
Example 4:

Input: "words and 987"
Output: 0
Explanation: The first non-whitespace character is 'w', which is not a numerical 
             digit or a +/- sign. Therefore no valid conversion could be performed.
Example 5:

Input: "-91283472332"
Output: -2147483648
Explanation: The number "-91283472332" is out of the range of a 32-bit signed integer.
             Thefore INT_MIN (−231) is returned.

"""

class Solution:
    def myAtoi(self, str):
        """
        :type str: str
        :rtype: int
        """
        i = num = 0
        while i<len(str) and str[i] == " ":
            i+=1
        if i>=len(str) or not str[i].isdigit() and str[i]!='+' and str[i]!='-':
            return 0
        sign = -1 if str[i] == '-' else 1
        if not str[i].isdigit():
            i+=1
        while i < len(str):
            if not str[i].isdigit(): 
                break
            num = num * 10 + int(str[i])
            i+=1
        num *= sign
        return 2147483647 if num > 2147483647 else -2147483648 if num <-2147483648 else num

class Solution(object):
    def myAtoi(self, s):

        ###better to do strip before sanity check (although 8ms slower):
        #ls = list(s.strip())
        #if len(ls) == 0 : return 0
        if len(s) == 0 : return 0
        ls = list(s.strip())
        
        sign = -1 if ls[0] == '-' else 1
        if ls[0] in ['-','+'] : del ls[0]
        ret, i = 0, 0
        while i < len(ls) and ls[i].isdigit() :
            ret = ret*10 + ord(ls[i]) - ord('0')
            i += 1
        return max(-2**31, min(sign * ret,2**31-1))



# Regex
class Solution:
    # @return an integer
    def atoi(self, str):
        str = str.strip()
        str = re.findall('(^[\+\-0]*\d+)\D*', str) # 开始以加号，减号 和的0开始的，后面接一个或多个digit
        #or   str = re.findall('^[+\-]?\d+', str)
        try:
            result = int(''.join(str))
            MAX_INT = 2147483647
            MIN_INT = -2147483648
            if result > MAX_INT > 0:
                return MAX_INT
            elif result < MIN_INT < 0:
                return MIN_INT
            else:
                return result
        except:
            return 0